feat: Added option to prevent attachment of individual files from knowledges

expruc opened the pull request:
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Verify that the pull request targets the `dev` branch. Not targeting the `dev` branch may lead to immediate closure of the PR.
- [x] **Description:** Provide a concise description of the changes made in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [ ] **Documentation:** If necessary, update relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs) like environment variables, the tutorials, or other documentation sources.
- [x] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [x] **Testing:** Perform manual tests to verify the implemented fix/feature works as intended AND does not break any other functionality. Take this as an opportunity to make screenshots of the feature/fix and include it in the PR description.
- [x] **Agentic AI Code:**: Confirm this Pull Request is **not written by any AI Agent** or has at least gone through additional human review **and** manual testing. If any AI Agent is the co-author of this PR, it may lead to immediate closure of the PR.
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Title Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

Added an option to prevent attachment of individual files from knowledges (only full knowledges are attachable, see screenshot below).

### Added

New environment variable `ENABLE_INDIVIDUAL_KNOWLEDGE_FILE_ATTACHMENTS` which controls whether attachment of single files from knowledges on the chat interface is allowed or not. If disabled, only full knowledges are allowed to be attached. Defaults to `true` (enabled).

### Additional Information

Motivation for this PR: slowness on cases with large knowledges (for example as mentioned in #17998). The cause for this slowness is loading all the knowledge in the background, files even if they are not attached or referenced by the knowledge.  

### Screenshots or Videos

<img width="480" height="270" alt="image" src="https://github.com/user-attachments/assets/fdfd1029-d46e-4a59-ae4e-cc695effa11d" />

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](https://github.com/open-webui/open-webui/blob/main/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

expruc committed:
Added option to prevent attachment of individual files from knowledges

expruc commented:
@deliciousbob I believe this is a cache images used as a part of the build process of the image, not the full image (you can check the [build workflow](https://github.com/expruc/open-webui/blob/main/.github/workflows/docker-build.yaml) for more information). You shouldn't use this tag as is, but other tags like the slim one https://github.com/expruc/open-webui/pkgs/container/open-webui/544274804?tag=feat-toggle_indivisual_files_attachment-slim, or basically any other tag which doesn't contain `cache` in it, according to your usecase.

I have tested the provided slim image locally using docker-compose and there were no problems, the application started normally and the new env variable worked as well.

deliciousbob commented:
Thx @expruc! This is the solution :) it workes perfectply :D 👍 
The list of Knolwedge-Collections is now instantly showing :) no problem with 10K files. Before it was like 7sec. waiting, now it is instant there! Thank you very much!

expruc commented:
Happy to hear that @deliciousbob thanks for your feedback!

expruc committed:
fixed knowledge attachment using # not filtering out knowledge files

expruc committed:
added option to toggle from UI (general settings)

expruc commented:
I have fixed the problem with the # (hashtag) knowledge references. Also added option to toggle this on the Admin UI (general settings).

<img width="1445" height="234" alt="image" src="https://github.com/user-attachments/assets/9eb21a8a-d9b9-44e5-985c-24b56a99f997" />

tjbck commented:
This cannot be merged as-is, if anything this should be refactored to create a new file table.

Classic298 commented:
I don't quite understand - a new file table? In addition to the already existing file table? Just to give admins the option to globally disable the ability to attach singular files when searching for a knowledgebase?

Maybe I am misunderstanding.

Either way, if a new table is introduced, the multitenancy setups for Qdrant and Milvus may need an update!

deliciousbob commented:
Hi @expruc! I've just imported your last Image to our Prod environment. Works well for me!
I now noticed that when you click onto the "+" Symbol to select the knowledge, it still loads for about 15 sec. until the Knowledge is shown in the list. This is happening on every page refresh. It needs to load that again. There is somehow a behavior for that that still loads mode data then necessarily neede for these tasks. 
Just testet it, it is the same load time on the "#" Hastag Knolwedge retrieval and "+" Option and when opening the Knowledge in the Worksacpe. They are tend to take about 15 sec. to load. 

It is still a substentially imporvement to having all files loaded, but there could still be some improvement done in the selection of that list from the DB.

tjbck commented:
@Classic298 Basically to map knowledge base <-> files, we should apply the same refactoring to our current user group handling logic.

pierrelouisbescond commented:
@expruc I had the same idea in June (https://github.com/open-webui/open-webui/discussions/15052).
Happy to see you were smart enough to implement that option 🤗
Looking forward to testing it when released!

expruc committed:
cleanup

expruc commented:
Any chance this gets reviewed now that #19278 has been merged? I have tested this PR after on v0.6.41 and still works. It is a different feature than the pagination (this doesn't change the logic of the files retrieval on the backend side).

tjbck commented:
This PR is likely not need after all the pref optimization refacs introduced in our latest dev branch. Confirmation wanted here!

expruc commented:
Yes @tjbck, this PR is redundant now after the recent changes in dev branch. Closing this PR, and I hope that v0.6.42 will be released soon! Thank you for all the hard work.
